merkle: pre‑reserve leaves to prevent reallocs with odd vtx count

l0rinc opened the pull request:
#### Summary

`ComputeMerkleRoot` [duplicates the last hash](https://github.com/bitcoin/bitcoin/blob/39b6c139bd6be33699af781f1d71f6fed303d468/src/consensus/merkle.cpp#L54-L56) when the input size is odd. If the caller provides a `std::vector` whose capacity equals its size, that extra `push_back` forces a reallocation, doubling its capacity (causing peak memory usage of 3x the necessary size).

This affects roughly half of the created blocks (those with odd transaction counts), causing unnecessary memory fragmentation during every block validation.

#### Fix

* Pre-reserves vector capacity to account for the odd-count duplication using `(size + 1) & ~1ULL`.
    * This syntax produces [optimal assembly](https://github.com/bitcoin/bitcoin/pull/32497#discussion_r2553107836) across x86/ARM and 32/64-bit platforms for GCC & Clang.
* Eliminates default construction of `uint256` objects that are immediately overwritten by switching from `resize` to `reserve` + `push_back`.

#### Memory Impact

[Memory profiling](https://github.com/bitcoin/bitcoin/pull/32497#issuecomment-3563724551) shows **50% reduction in peak allocation** (576KB → 288KB) and elimination of reallocation overhead.

#### Validation

The benchmark was updated to use an odd leaf count to demonstrate the real-world scenario where the reallocation occurs.

A full `-reindex-chainstate` up to block **896 408** ran without triggering the asserts.

<details>
<summary>Validation asserts</summary>

Temporary asserts (not included in this PR) confirm that `push_back` never reallocates and that the coinbase witness hash remains null:
```cpp
if (hashes.size() & 1) {
    assert(hashes.size() < hashes.capacity()); // TODO remove
    hashes.push_back(hashes.back());
}

leaves.reserve((block.vtx.size() + 1) & ~1ULL); // capacity rounded up to even
leaves.emplace_back();
assert(leaves.back().IsNull()); // TODO remove
```

</details>

#### Benchmark Performance

While the main purpose is to improve predictability, the reduced memory operations also improve hashing throughput slightly.

l0rinc committed:
bench: make `MerkleRoot` benchmark more representative

Two versions are run now, one with the mutation calculations, the other without.
To avoid unwanted compiler optimizations, we assert the expected hash, which should inhibit aggressive optimization.

To make the benchmark more similar to production `ComputeMerkleRoot` call sites, the input leaves-copying is made explicit before each run.

> ./build/bin/bench_bitcoin -filter='MerkleRoot.*' -min-time=1000

|             ns/leaf |              leaf/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               44.18 |       22,634,858.70 |    0.0% |      1.10 | `MerkleRoot`
|               44.66 |       22,390,601.03 |    0.0% |      1.10 | `MerkleRootWithMutation`

Massif memory measurements show the excessive memory reservations:

    MB
1.332^        :
     | #      :
     | #      :
     | #      :
     | #      :
     | #    @ :
     | #    @ :
     | #    @ :
     | #    @ :
     | #    @ :
     | #    @ :
     | #    @ :
     | #    @ :
     | #::::@::::::::::::::::::::::::::::::::::::::::::::::::::::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
   0 +----------------------------------------------------------------------->s
     0                                                                   226.2

showing the reallocations clearly in the stacks:
97.87% (1,366,841B) (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
->41.25% (576,064B) 0x969717: allocate (new_allocator.h:151)
| ->41.25% (576,064B) 0x969717: allocate (allocator.h:203)
|   ->41.25% (576,064B) 0x969717: allocate (alloc_traits.h:614)
|     ->41.25% (576,064B) 0x969717: _M_allocate (stl_vector.h:387)
|       ->41.25% (576,064B) 0x969717: _M_realloc_append<const uint256&> (vector.tcc:572)
|         ->41.25% (576,064B) 0x969717: push_back (stl_vector.h:1427)
|           ->41.25% (576,064B) 0x969717: ComputeMerkleRoot(std::vector<uint256, std::allocator<uint256> >, bool*) (merkle.cpp:55)
|             ->41.25% (576,064B) 0x2235A7: operator() (merkle_root.cpp:31)
|               ->41.25% (576,064B) 0x2235A7: ankerl::nanobench::Bench& ankerl::nanobench::Bench::run<MerkleRoot(ankerl::nanobench::Bench&)::{lambda()

Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

luke-jr commented on src/consensus/merkle.cpp:
```diff
@@ -66,20 +66,20 @@ uint256 ComputeMerkleRoot(std::vector<uint256> hashes, bool* mutated) {
-    leaves.resize(block.vtx.size());
-    for (size_t s = 0; s < block.vtx.size(); s++) {
-        leaves[s] = block.vtx[s]->GetHash();
+    leaves.reserve((block.vtx.size() + 1) & ~1ULL); // capacity rounded up to even
```
This assumes `size_t` is `unsigned long long` I think? Can we avoid that easily?

l0rinc replied:
we could either do `size_t{1}` or `leaves.reserve(block.vtx.size() + (block.vtx.size() & 1));`, I think, if you think this could be problematic on 32 bit systems.

l0rinc replied:
Thanks for checking, I think it's the best solution regarding C++ simplicity and CPU support.

luke-jr commented on src/bench/merkle_root.cpp:
```diff
@@ -11,16 +11,15 @@
-    std::vector<uint256> leaves;
-    leaves.resize(9001);
+    FastRandomContext rng{/*fDeterministic=*/true};
+    std::vector<uint256> leaves(9000); // even number of leaves
```
Wouldn't it be better to benchmark the less-performant input?

l0rinc replied:
After this change all call-sites are pre-sized to even number of leaves, so this reflects the usage better.
I could also round to `9002` to mimic the new `leaves.reserve((9001 + 1) & ~1ULL)` behavior, if you thinks that's more representative.

luke-jr commented on src/bench/merkle_root.cpp:
```diff
@@ -11,16 +11,15 @@
-        leaves[mutation] = hash;
+        bool mutation{false};
+        const uint256 hash{ComputeMerkleRoot(std::vector(leaves), &mutation)};
+        assert(mutation == false && hash == uint256{"18ff9b2bef46a834b2625871b2839fec3b67fc1243920c424858324fbde1a8de"});
```
How does this avoid optimisations?

l0rinc replied:
the `ComputeMerkleRoot` calculation has to be performed to be able to validate this assert - it's can't eliminate the call.

luke-jr reviewed (approved):
utACK main commit. Not sure why the benchmark is being changed.

l0rinc reviewed (commented):
Thanks for the review!

yuvicc reviewed (commented):
Concept ACK

Memory allocation/deallocation is slow and expensive.

yuvicc commented:
I have thought on changing the benchmark tests as we would want to test the worst case scenario right? or else we could have both case for testing?

yuvicc reviewed (commented):
# Benchmark Testing

master@5757de4ddd37f9321ee6b338b40888fd3561fc00

- With 9000
```
|             ns/leaf |              leaf/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               52.62 |       19,005,746.07 |    0.2% |      0.01 | `MerkleRoot`
|               52.64 |       18,998,504.40 |    0.3% |      0.01 | `MerkleRoot`
|               52.63 |       18,999,727.67 |    0.2% |      0.01 | `MerkleRoot`
```

- With 9001
```
|             ns/leaf |              leaf/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               53.50 |       18,693,063.88 |    0.3% |      0.01 | `MerkleRoot`
|               53.53 |       18,681,211.49 |    0.5% |      0.01 | `MerkleRoot`
|               53.49 |       18,694,053.87 |    0.5% |      0.01 | `MerkleRoot`
```

--------------------------------------------------------------------

Commit 39b6c139bd6be33699af781f1d71f6fed303d468

- With 9000
```
|             ns/leaf |              leaf/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               52.51 |       19,043,628.95 |    0.2% |      0.01 | `MerkleRoot`
|               52.52 |       19,040,989.96 |    0.2% |      0.01 | `MerkleRoot`
|               52.53 |       19,036,358.39 |    0.2% |      0.01 | `MerkleRoot`
```

- With 9001
```
|             ns/leaf |              leaf/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               53.40 |       18,713,525.67 |    0.3% |      0.01 | `MerkleRoot`
|               53.44 |       19,314,655.73 |    0.3% |      0.01 | `MerkleRoot`
|               53.41 |       18,883,462.75 |    0.3% |      0.01 | `MerkleRoot`
```

l0rinc commented:
Update the benchmark to showcase the before/after state better (resembles production code changes), by splitting out the vector copies to the unmetered lambda.
Changed the previous `leaves.reserve((block.vtx.size() + 1) & ~1ULL)` rounding to `leaves.reserve(block.vtx.size() + (block.vtx.size() & 1))`.

yuvicc reviewed (commented):
Code review ACK 4cc5942895673591de4edceb6dd0c7188c302a72

# Testing

## Before Master@9a7eece5a4
|             ns/leaf |              leaf/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               53.75 |       18,605,160.70 |    0.7% |      0.01 | `MerkleRoot`
|               53.64 |       18,643,002.35 |    0.1% |      0.01 | `MerkleRoot`
|               53.91 |       18,548,704.52 |    0.4% |      0.01 | `MerkleRoot`

## After commit@4cc5942895673591de4edceb6dd0c7188c302a72
|             ns/leaf |              leaf/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               57.67 |       17,340,305.39 |    0.0% |      0.52 | `MerkleRoot`
|               57.69 |       17,332,534.96 |    0.0% |      0.52 | `MerkleRoot`
|               58.14 |       17,199,057.10 |    0.0% |      0.52 | `MerkleRoot`

maflcko reviewed (commented):
not sure how meaningful this is, but it looks like there are a bunch of unrelated style changes that don't seem to have a benefit?

l0rinc reviewed (commented):
The changes revert the original benchmark's behavior of including the leaf vector in the benchmark (more explicitly and reserved like the production code).
I've also changed the `emplace_back` calls for the hashes to `push_bash` and reverted the original loop conditions to minimize the diff in critical code sections.

Note to yuvicc: previous benchmark version was measuring something else than master (only the merkle call without the vector copy), so we can't directly compare the results.

~The new results don't show any significant speed increase, only the memory allocation is more predictable - which isn't captured by the benchmark anymore.~

l0rinc commented:
I've rebased the changed and adjusted the benchmark to be more similar to the other production usages in the first commit, rounded to even in the second (optimization) commit, so that we can realistically measure the speed difference before and after:

% build/bin/bench_bitcoin -filter='MerkleRoot' --min-time=1000

Before 7f620cffebee593e48434cf182cc2fd64a6d76be:

|             ns/leaf |              leaf/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               45.50 |       21,975,688.66 |    0.0% |      1.10 | `MerkleRoot`
|               45.53 |       21,962,546.91 |    0.1% |      1.10 | `MerkleRoot`
|               45.53 |       21,965,162.52 |    0.0% |      1.10 | `MerkleRoot`

After a73dd45e95f5e909d0950ba7c76225589df74854:
|             ns/leaf |              leaf/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               45.04 |       22,200,976.95 |    0.1% |      1.10 | `MerkleRoot`
|               44.97 |       22,235,208.25 |    0.1% |      1.10 | `MerkleRoot`
|               45.01 |       22,217,033.81 |    0.1% |      1.10 | `MerkleRoot`

l0rinc commented:
Rebased and added a new commit on top for deduplicating the integer rounding. Ready for review again.

optout21 commented:
LGTM!
Minor comments:
- Maybe add a note to the `ComputeMerkleRoot` declaration, that the `hashes` input is expected to have one extra capacity (in case size is odd). Minor, as this function is not used directly much.
- Even more, it could debug-assert that capacity is even, but this may be a bit of a stretch.

optout21 reviewed (commented):
reACK 3dd815f048c80c9a35f01972e0537eb42531aec7
Change since last review: Ajdust a leftover assert in the test code.

(prev:
reACK 8af0a3a72c8af781bebed7289657c637220df893
ACK e3b02f1539e59a6a20a00b4161219b5aba6fefed
ReACK. Nicely simplified.
a7997d02ebe5b72e706712f43884644b705a1cd1  Good optimization, nice catch!)

achow101 commented on src/test/fuzz/integer.cpp:
```diff
@@ -80,7 +81,8 @@ FUZZ_TARGET(integer, .init = initialize_integer)
     constexpr uint256 u256_max{"ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"};
-    const std::vector<uint256> v256{u256, u256_min, u256_max};
+    std::vector<uint256> v256{u256, u256_min, u256_max};
+    v256.reserve(RoundUpToEven(v256.size()));
```
This reserve seems to be for `ComputeMerkleRoot`. However, I think it would be better to have `ComputeMerkleRoot` do any reservations necessary rather than relying on the caller to do so.

l0rinc replied:
`ComputeMerkleRoot` already received a list of values, usually sized to fit their current size exactly (i.e. `capacity()` == `size()`). We can of course increase the capacity inside - it's what:
```C++
if (hashes.size() & 1) {
    hashes.push_back(hashes.back());
}
```
does already, but if we want to avoid reallocations, we have to predict the size of the vector before we start populating it.

-----

We could add some kind of lambda as a parameter and some starting values (an expected size and a generator function maybe), something like:
```C++
template <typename Gen>
uint256 ComputeMerkleRoot2(std::size_t count, Gen&& gen, bool* mutated)
{
    std::vector<uint256> leaves;
    leaves.reserve(RoundUpToEven(count));
    for (std::size_t i{0}; i < count; ++i) {
        leaves.push_back(gen(i));
    }
    return ComputeMerkleRoot(std::move(leaves), mutated);
}
```
and the production code callers would look like:
```C++
static uint256 ComputeModifiedMerkleRoot(const CMutableTransaction& cb, const CBlock& block)
{
    return ComputeMerkleRoot2(block.vtx.size(), [&](auto i) { return i == 0 ? cb.GetHash() : block.vtx[i]->GetHash(); });
}

uint256 BlockMerkleRoot(const CBlock& block, bool* mutated)
{
    return ComputeMerkleRoot2(block.vtx.size(), [&](auto i) { return block.vtx[i]->GetHash(); }, mutated);
}
uint256 BlockWitnessMerkleRoot(const CBlock& block, bool* mutated)
{
    return ComputeMerkleRoot2(block.vtx.size(), [&](auto i) { return i == 0 ? uint256{} : uint256(block.vtx[i]->GetWitnessHash()); }, mutated);
}
```

Would that be simpler in your opinion?

<details>
<summary>Patch</summary>

```patch
diff --git a/src/consensus/merkle.cpp b/src/consensus/merkle.cpp
index 8c56d979a5..a0e358b005 100644
--- a/src/consensus/merkle.cpp
+++ b/src/consensus/merkle.cpp
@@ -5,7 +5,6 @@
 #include <consensus/merkle.h>
 #include <hash.h>
 #include <util/check.h>
-#include <util/ints.h>
 
 /*     WARNING! If you're reading this because you're learning about crypto
        and/or designing a new system that will use merkle trees, keep in mind
@@ -63,26 +62,14 @@ uint256 ComputeMerkleRoot(std::vector<uint256> hashes, bool* mutated) {
     return hashes[0];
 }
 
-
 uint256 BlockMerkleRoot(const CBlock& block, bool* mutated)
 {
-    std::vector<uint256> leaves;
-    leaves.reserve(RoundUpToEven(block.vtx.size()));
-    for (size_t s = 0; s < block.vtx.size(); s++) {
-        leaves.push_back(block.vtx[s]->GetHash());
-    }
-    return ComputeMerkleRoot(std::move(leaves), mutated);
+    return GenerateMerkleRoot(block.vtx.size(), [&](auto i) { return block.vtx[i]->GetHash(); }, mutated);
 }
 
 uint256 BlockWitnessMerkleRoot(const CBlock& block, bool* mutated)
 {
-    std::vector<uint256> leaves;
-    leaves.reserve(RoundUpToEven(block.vtx.size()));
-    leaves.emplace_back(); // The witness hash of the coinbase is 0.
-    for (size_t s = 1; s < block.vtx.size(); s++) {
-        leaves.push_back(block.vtx[s]->GetWitnessHash());
-    }
-    return ComputeMerkleRoot(std::move(leaves), mutated);
+    return GenerateMerkleRoot(block.vtx.size(), [&](auto i) { return i == 0 ? uint256{} : uint256(block.vtx[i]->GetWitnessHash()); }, mutated);
 }
 
 /* This implements a constant-space merkle root/path calculator, limited to 2^32 leaves. */
diff --git a/src/consensus/merkle.h b/src/consensus/merkle.h
index c722cbe446..8b5c14157a 100644
--- a/src/consensus/merkle.h
+++ b/src/consensus/merkle.h
@@ -9,9 +9,21 @@
 
 #include <primitives/block.h>
 #include <uint256.h>
+#include <util/ints.h>
 
 uint256 ComputeMerkleRoot(std::vector<uint256> hashes, bool* mutated = nullptr);
 
+template <typename Gen>
+uint256 GenerateMerkleRoot(std::size_t count, Gen&& gen, bool* mutated = nullptr)
+{
+    std::vector<uint256> leaves;
+    leaves.reserve(RoundUpToEven(count));
+    for (std::size_t i{0}; i < count; ++i) {
+        leaves.push_back(gen(i));
+    }
+    return ComputeMerkleRoot(std::move(leaves), mutated);
+}
+
 /*
  * Compute the Merkle root of the transactions in a block.
  * *mutated is set to true if a duplicated subtree was found.
diff --git a/src/signet.cpp b/src/signet.cpp
index 3104b7f2d7..5aea084f15 100644
--- a/src/signet.cpp
+++ b/src/signet.cpp
@@ -58,13 +58,8 @@ static bool FetchAndClearCommitmentSection(const std::span<const uint8_t> header
 
 static uint256 ComputeModifiedMerkleRoot(const CMutableTransaction& cb, const CBlock& block)
 {
-    std::vector<uint256> leaves;
-    leaves.reserve(RoundUpToEven(block.vtx.size()));
-    leaves.push_back(cb.GetHash());
-    for (size_t s = 1; s < block.vtx.size(); ++s) {
-        leaves.push_back(block.vtx[s]->GetHash());
-    }
-    return ComputeMerkleRoot(std::move(leaves));
+    return GenerateMerkleRoot(block.vtx.size(),
+        [&](auto i) { return i ? block.vtx[i]->GetHash() : cb.GetHash(); });
 }
 
 std::optional<SignetTxs> SignetTxs::Create(const CBlock& block, const CScript& challenge)
```

</details>

achow101 replied:
> Would that be simpler in your opinion?

No.

> but if we want to avoid reallocations, we have to predict the size of the vector before we start populating it.

We can also avoid inserting into the vector in the first place, e.g. this diff gives me about the same speedup without needing to change anything with allocations:

```diff
diff --git a/src/consensus/merkle.cpp b/src/consensus/merkle.cpp
index 7dd24e1868f..46056bfa3e0 100644
--- a/src/consensus/merkle.cpp
+++ b/src/consensus/merkle.cpp
@@ -44,6 +44,7 @@
 
 
 uint256 ComputeMerkleRoot(std::vector<uint256> hashes, bool* mutated) {
+    std::array<uint256, 2> odd_hash;
     bool mutation = false;
     while (hashes.size() > 1) {
         if (mutated) {
@@ -51,11 +52,20 @@ uint256 ComputeMerkleRoot(std::vector<uint256> hashes, bool* mutated) {
                 if (hashes[pos] == hashes[pos + 1]) mutation = true;
             }
         }
-        if (hashes.size() & 1) {
+        bool need_dup = hashes.capacity() == hashes.size() && (hashes.size() & 1);
+        if (need_dup) {
+            odd_hash[0] = hashes.back();
+            odd_hash[1] = hashes.back();
+        } else if (hashes.size() & 1) {
             hashes.push_back(hashes.back());
         }
         SHA256D64(hashes[0].begin(), hashes[0].begin(), hashes.size() / 2);
         hashes.resize(hashes.size() / 2);
+        if (need_dup) {
+            SHA256D64(odd_hash[0].begin(), odd_hash[0].begin(), 1);
+            assert(hashes.capacity() > hashes.size());
+            hashes.push_back(odd_hash[0]);
+        }
     }
     if (mutated) *mutated = mutation;
     if (hashes.size() == 0) return uint256();
```

l0rinc replied:
Clever, we can of course calculate the extra value independently, but isn't it a lot simpler to just precalculate the final size before insertions and keep `ComputeMerkleRoot` unchanged?

achow101 replied:
I think it's simpler if the caller does not have to be aware that they should allocate a larger vector to be passed in.

l0rinc replied:
The lambda solution is even simpler in that case, the caller doesn't even have to allocate or iterate at all.

l0rinc replied:
Addressed it in a different way in latest push, let me know if this solves your concern or if it's orthogonal.

l0rinc commented:
> Unless observable benefits can be shown

The change itself is trivial, it's just a **+1** for the reserve before calling the Merkle calculation.
I don't mind simplifying the change if you think that's easier to review (I could also just bump `leaves.reserve(block.vtx.size())` to `leaves.reserve(block.vtx.size() + 1)` without the refactors).
Originally it was similar to that, but based on [concerns](https://github.com/bitcoin/bitcoin/pull/32497#discussion_r2264185044) to decouple callers from needing to know the internals of Merkle calculations, I extracted the loops.

I'm open to simplifying it back to just a reserve if you think that's easier to review, but I'd argue that the underlying incorrect memory reservation is a minor bug that should be fixed.

---------

While this change does result in small speedup, the primary purpose is to avoid having to reallocate 2x additional memory for a predictable calculation - less memory fragmentation, lower memory usage, more predictable behavior.
Let's quantify the impact: running `massif` for the benchmark with an odd leaf count shows the memory peaks clearly:

<details>
<summary>massif measurement comparing the two commits for 200 second benchmark run</summary>

```bash
COMMITS="f2334a0ae056a9e3cba218206d139b3688eeda96 de6a28396b8bf7a3fd1124418b4cd5beba1906d6"; \
CC=gcc; CXX=g++; \
BASE_DIR="/mnt/my_storage"; LOG_DIR="$BASE_DIR/logs"; \
FILTER="MerkleRoot"; MIN_TIME=200000; \
mkdir -p "$LOG_DIR"; \
(echo "memory benchmark | filter=$FILTER | min-time=${MIN_TIME}ms | $(hostname) | $(uname -m) | $(lscpu | grep 'Model name' | head -1 | cut -d: -f2 | xargs) | $(nproc) cores | $(free -h | awk '/^Mem:/{print $2}') RAM";) &&\
(for c in $COMMITS; do git log -1 --pretty='%h %s' $c || exit 1; done; echo "") && \
for COMMIT in $COMMITS; do \
  SHORT_COMMIT=${COMMIT:0:8}; \
  echo "Processing commit $SHORT_COMMIT..."; \
  git fetch -q origin $COMMIT && git checkout -q $COMMIT && \
  git clean -fxd && git reset --hard && \
  cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUILD_BENCH=ON -DENABLE_IPC=OFF && \
  ninja -C build bench_bitcoin -j2 && \
  valgrind --tool=massif --time-unit=ms \
    --massif-out-file="$LOG_DIR/massif-bench-$FILTER-$SHORT_COMMIT.out" \
    build/bin/bench_bitcoin -filter="$FILTER" --min-time=$MIN_TIME && \
  ms_print "$LOG_DIR/massif-bench-$FILTER-$SHORT_COMMIT.out" | tee "$LOG_DIR/massif-bench-$FILTER-$SHORT_COMMIT.txt" && \
  echo "Completed $SHORT_COMMIT"; \
done && \
echo "" && echo "Results in: $LOG_DIR/massif-bench-$FILTER-*.{out,txt}"
````

</details>

We can also clearly see the [difference](https://www.diffchecker.com/Lrp4SYcS) in the capture stack traces.
**Before**, having  **1.3 MB peak memory** usage:
```massif
    MB
1.332^        :                                                               
     | #      :                                                               
     | #      :                                                               
     | #      :                                                               
     | #      :                                                               
     | #    @ :                                                               
     | #    @ :                                                               
     | #    @ :                                                               
     | #    @ :                                                               
     | #    @ :                                                               
     | #    @ :                                                               
     | #    @ :                                                               
     | #    @ :                                                               
     | #::::@::::::::::::::::::::::::::::::::::::::::::::::::::::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
     | #: ::@::::: :::::::: :: ::: :::::: : : :: ::: ::: : : : ::::@:::::@::::
   0 +----------------------------------------------------------------------->s
     0                                                                   226.2
```
showing the reallocations clearly in the stacks:
```
97.87% (1,366,841B) (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
->41.25% (576,064B) 0x969717: allocate (new_allocator.h:151)
| ->41.25% (576,064B) 0x969717: allocate (allocator.h:203)
|   ->41.25% (576,064B) 0x969717: allocate (alloc_traits.h:614)
|     ->41.25% (576,064B) 0x969717: _M_allocate (stl_vector.h:387)
|       ->41.25% (576,064B) 0x969717: _M_realloc_append<const uint256&> (vector.tcc:572)
|         ->41.25% (576,064B) 0x969717: push_back (stl_vector.h:1427)
|           ->41.25% (576,064B) 0x969717: ComputeMerkleRoot(std::vector<uint256, std::allocator<uint256> >, bool*) (merkle.cpp:55)
|             ->41.25% (576,064B) 0x2235A7: operator() (merkle_root.cpp:31)
|               ->41.25% (576,064B) 0x2235A7: ankerl::nanobench::Bench& ankerl::nanobench::Bench::run<MerkleRoot(ankerl::nanobench::Bench&)::{lambda()
```

-----

**After**, having a **0.8 MB peak memory** usage
```massif
    KB
801.4^ #                                                                      
     | #                                                                      
     | #                                                                      
     | #                                                                      
     | #                                                                      
     | #                                                                      
     | #                                                                      
     | #                                                         :::::@:::::@:
     | #:::@@@::@:::::::::::::::@::@:@:::@@:::::::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
   0 +----------------------------------------------------------------------->s
     0                                                                   227.5
```
and the stacks don't show reallocs anymore:
```
96.37% (790,809B) (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
->35.10% (288,064B) 0x2234AF: allocate (new_allocator.h:151)
| ->35.10% (288,064B) 0x2234AF: allocate (allocator.h:203)
|   ->35.10% (288,064B) 0x2234AF: allocate (alloc_traits.h:614)
|     ->35.10% (288,064B) 0x2234AF: _M_allocate (stl_vector.h:387)
|       ->35.10% (288,064B) 0x2234AF: reserve (vector.tcc:79)
|         ->35.10% (288,064B) 0x2234AF: ToMerkleLeaves<std::vector<uint256>, MerkleRoot(ankerl::nanobench::Bench&)::<lambda()>::<lambda(bool, const auto:46&)> > (merkle.h:19)
|           ->35.10% (288,064B) 0x2234AF: operator() (merkle_root.cpp:25)
|             ->35.10% (288,064B) 0x2234AF: ankerl::nanobench::Bench& ankerl::nanobench::Bench::run<MerkleRoot(ankerl::nanobench::Bench&)::{lambda()
```

hodlinator reviewed (commented):
Concept ACK de6a28396b8bf7a3fd1124418b4cd5beba1906d6

`ComputeMerkleRoot(std::vector<uint256> hashes, ...)` takes a vector by value, but luckily call sites already move into it, and since the underlying array/memory is transferred on move, the capacity "transfers as well" - neat!

---

> Sorry to be this guy again, but it seems to me unreasonable to touch such critical consensus code for such marginal benefits. Unless observable benefits can be shown, ...

Thanks for being that guy!

Going by https://github.com/bitcoin/bitcoin/pull/32497#issuecomment-3006209711 and https://github.com/bitcoin/bitcoin/pull/32497#issuecomment-3011987044 there seems to be a 2% speed increase for computing merkle roots.

I cannot spot any issue with how this PR currently changes the consensus code. However, maybe we could reduce the scope of the change somewhat to make it more palatable. Here is my attempt:
https://github.com/bitcoin/bitcoin/compare/master...hodlinator:bitcoin:refs/heads/pr/32497_minimal#diff-706988c23877f8a557484053887f932b2cafb3b5998b50497ce7ff8118ac85a3 (link directly to src/consensus/merkle.cpp)

This 2% speed increase in merkle root computation is probably not at all measurable for a full IBD though. So if there is no variant of this optimization that inspires confidence, I can understand the push-back.

l0rinc committed:
test: adjust `ComputeMerkleRoot` tests

Update the integer fuzz test to move the vector into `ComputeMerkleRoot`, matching production usage patterns and avoiding unnecessary copies.

Update `merkle_test_BlockWitness` to use an odd number of transactions to ensure the test covers the scenario where leaf duplication occurs. Also switch to `GetWitnessHash` to match `BlockWitnessMerkleRoot` semantics.
The manual vector setup retains the exact-size `resize` to explicitly verify the behavior against the calculated root.

l0rinc committed:
validation: pre-reserve leaves to prevent reallocs with odd vtx count

`ComputeMerkleRoot` duplicates the last hash when the input size is odd. If the caller provides a `std::vector` whose capacity equals its size, that extra `push_back` forces a reallocation, doubling its capacity (allocating 3x the necessary memory).

This affects roughly half of the created blocks (those with odd transaction counts), causing unnecessary memory fragmentation during every block validation.

Fix this by pre-reserving the vector capacity to account for the odd-count duplication. The expression `(size + 1) & ~1ULL` adds 1 to the size and clears the last bit, effectively rounding up to the next even number. This syntax produces optimal assembly across x86/ARM and 32/64-bit platforms for gcc/clang, see https://godbolt.org/z/xzscoq7nv.

Also switch from `resize` to `reserve` + `push_back` to eliminate the default construction of `uint256` objects that are immediately overwritten.

> ./build/bin/bench_bitcoin -filter='MerkleRoot.*' -min-time=1000

|             ns/leaf |              leaf/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               43.73 |       22,867,350.51 |    0.0% |      1.10 | `MerkleRoot`
|               44.17 |       22,640,349.14 |    0.0% |      1.10 | `MerkleRootWithMutation`

Massif memory measurements after show 0.8 MB peak memory usage

    KB
801.4^ #
     | #
     | #
     | #
     | #
     | #
     | #
     | #                                                         :::::@:::::@:
     | #:::@@@::@:::::::::::::::@::@:@:::@@:::::::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
     | #:::@ @: @:::::::::::::::@::@:@:::@ :::: ::::@::::::@:::::@::::@:::::@:
   0 +----------------------------------------------------------------------->s
     0                                                                   227.5

and the stacks don't show reallocs anymore:
96.37% (790,809B) (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
->35.10% (288,064B) 0x2234AF: allocate (new_allocator.h:151)
| ->35.10% (288,064B) 0x2234AF: allocate (allocator.h:203)
|   ->35.10% (288,064B) 0x2234AF: allocate (alloc_traits.h:614)
|     ->35.10% (288,064B) 0x2234AF: _M_allocate (stl_vector.h:387)
|       ->35.10% (288,064B) 0x2234AF: reserve (vector.tcc:79)
|         ->35.10% (288,064B) 0x2234AF: ToMerkleLeaves<std::vector<uint256>, MerkleRoot(ankerl::nanobench::Bench&)::<lambda()>::<lambda(bool, const auto:46&)> > (merkle.h:19)
|           ->35.10% (288,064B) 0x2234AF: operator() (merkle_root.cpp:25)
|             ->35.10% (288,064B) 0x2234AF: ankerl::nanobench::Bench& ankerl::nanobench::Bench::run<MerkleRoot(ankerl::nanobench::Bench&)::{lambda()

Co-authored-by: optout21 <13562139+optout21@users.noreply.github.com>
Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

l0rinc reviewed (commented):
Latest push:
* rebases against latest master;
* reverts code back to simply change `leaves.resize(block.vtx.size())` constructs before calling `ComputeMerkleRoot` to `.reserve((block.vtx.size() + 1) & ~1ULL)`;
  ** added godbolt reproducer for each important combination of compiler and architecture
* Separates the test changes to a separate commit explaining the motivation in the commit message
* updated commit messages and PR descriptions.

optout21 commented on src/bench/merkle_root.cpp:
```diff
@@ -7,20 +7,29 @@
+    bench.batch(hashes.size()).unit("leaf").run([&] {
+        std::vector<uint256> leaves;
+        leaves.reserve(hashes.size());
+        for (size_t s = 0; s < hashes.size(); s++) {
```
nit: consistent coding style
```diff
-        for (size_t s = 0; s < hashes.size(); s++) {
+        for (auto s{0}; s < hashes.size(); ++s) {
```

optout21 commented on src/bench/merkle_root.cpp:
```diff
@@ -7,20 +7,29 @@
+        leaves.reserve(hashes.size());
+        for (size_t s = 0; s < hashes.size(); s++) {
+            leaves.push_back(hashes[s]);
+        }
```
nit: Could this copy be done simpler? (i.e. `leaves = hashes`, or `leaves{hashes}`)

hodlinator reviewed (commented):
Reviewed e3b02f1539e59a6a20a00b4161219b5aba6fefed

Thanks for reverting to a simpler version of this change!

2 inline questions.

l0rinc commented on src/bench/merkle_root.cpp:
```diff
@@ -7,20 +7,29 @@
     }
-    bench.batch(leaves.size()).unit("leaf").run([&] {
-        bool mutation = false;
-        uint256 hash = ComputeMerkleRoot(std::vector<uint256>(leaves), &mutation);
```
My reasoning was that I want the benchmark to reflect the actual usage, see: https://github.com/bitcoin/bitcoin/pull/32497#issuecomment-2973633245

It wouldn't be fair to see the `ComputeMerkleRoot` performance without the leaf constructions (and get an unrealistic speedup). What do you think?

l0rinc replied:
`ComputeMerkleRoot` is usually called without mutation checks:
https://github.com/bitcoin/bitcoin/blob/fa283d28e261416f0785450ab687af199103776a/src/validation.cpp#L3929-L3932

Most existing mutation checks are noops, see: https://github.com/bitcoin/bitcoin/pull/33805.
The benchmark was modeling pre-segwit `BlockMerkleRoot` before, and now it's closer to the behavior of `BlockWitnessMerkleRoot`.

I haven't measured it explicitly, but the latter should be more representative of modern usage - please let me know if you think I misunderstood something.

l0rinc replied:
I can also just revert the benchmark completely, this is mostly a correctness (reserve the actual amount, not one less) and memory fix (don't allocate 3x peak memory), I don't expect it to speed up execution a lot, but I do expect it to fragment the memory less, see https://github.com/bitcoin/bitcoin/pull/32497#issuecomment-3563724551

hodlinator replied:
I prefer doing the reserve-adjustment in the benchmark as well and doing the reserve inside of the benchmarked section (edited https://github.com/bitcoin/bitcoin/pull/32497#discussion_r2569912274 earlier today).

Makes sense that it could have a higher impact on memory fragmentation than execution speed.

hodlinator reviewed (approved):
ACK 8af0a3a72c8af781bebed7289657c637220df893

While the benefit of this change to consensus code is low, it's been simplified back to a more easily reviewable version.

The benchmark now has two parts, one that detects mutation issues and one which does not, corresponding to the non-witness & witness merkle root computations performed by Bitcoin Core on mainnet.

### One extra level of paranoia?

I think this change really is simple enough to approve as-is. It's maybe not the same risk-level as changing standard library major versions, but adjacent to it. If people still object, one idea to improve confidence could be to splice in a new commit before the last one - it could add `BlockMerkleRootNew()` and `BlockWitnessMerkleRootNew()` alongside the old ones, and have a fuzz test which compares results between new & old versions. Then the last commit could merge new & old together, keeping the old name but using the new implementations, and removing the fuzz test again.

### Benchmarks

<details><summary>Expand</summary>

```shell
₿ cmake --build build -t bench_bitcoin && build/bin/bench_bitcoin -filter="MerkleRoot" -min-time=10000
```

#### 5fd62ede95900a6ea603c2b9f14c58ff4c32a4e2 "bench: make `MerkleRoot` benchmark more representative"

```
|             ns/leaf |              leaf/s |    err% |        ins/leaf |        cyc/leaf |    IPC |       bra/leaf |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|              115.53 |        8,655,444.89 |    0.0% |        1,216.23 |          425.94 |  2.855 |           2.14 |    0.3% |     11.03 | `MerkleRoot`
|              116.84 |        8,558,882.10 |    0.0% |        1,231.24 |          430.78 |  2.858 |           3.14 |    0.2% |     10.99 | `MerkleRootWithMutation`
```

#### 8af0a3a72c8af781bebed7289657c637220df893 "validation: pre-reserve leaves to prevent reallocs with odd vtx count"

```
|             ns/leaf |              leaf/s |    err% |        ins/leaf |        cyc/leaf |    IPC |       bra/leaf |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|              113.25 |        8,830,206.60 |    0.0% |        1,223.17 |          417.50 |  2.930 |           2.63 |    0.2% |     10.99 | `MerkleRoot`
|              114.67 |        8,720,906.94 |    0.0% |        1,238.18 |          422.72 |  2.929 |           3.63 |    0.2% |     11.01 | `MerkleRootWithMutation`
```

#### Modified 8af0a3a72c8af781bebed7289657c637220df893 which just does `leaves.reserve(hashes.size())`

```
|             ns/leaf |              leaf/s |    err% |        ins/leaf |        cyc/leaf |    IPC |       bra/leaf |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|              114.81 |        8,709,871.88 |    0.0% |        1,230.23 |          423.26 |  2.907 |           3.64 |    0.2% |     11.01 | `MerkleRoot`
|              116.21 |        8,605,021.23 |    0.0% |        1,245.23 |          428.41 |  2.907 |           4.64 |    0.2% |     11.00 | `MerkleRootWithMutation`
```

This last one shows that even just switching from `resize()` to `reserve()`, without always reserving an even number, does result in a measurable speedup by itself. Explains why I didn't see as big differences in my previous testing (https://github.com/bitcoin/bitcoin/pull/32497#discussion_r2576533841) since all those variants used `reserve()`.

</details>

vasild commented on src/test/merkle_tests.cpp:
```diff
@@ -237,12 +238,12 @@ BOOST_AUTO_TEST_CASE(merkle_test_BlockWitness)
-    hashes[0].SetNull();
-    hashes[1] = block.vtx[1]->GetHash().ToUint256();
+    hashes.resize(vtx_count); // Note: leaving odd count to exercise old behavior
+    for(size_t pos{1}; pos < vtx_count; pos++) {
```
style: space after `for` and `++i` is preferred over `i++` (from `doc/developer-notes.md`).
```suggestion
    for (size_t pos{1}; pos < vtx_count; ++pos) {
```
(the loop a few lines above also uses `pos++`)

vasild commented on src/bench/merkle_root.cpp:
```diff
@@ -7,21 +7,34 @@
+    for (bool mutate : {false, true}) {
+        bench.name(mutate ? "MerkleRootWithMutation" : "MerkleRoot").batch(hashes.size()).unit("leaf").run([&] {
+            std::vector<uint256> leaves;
+            leaves.reserve((hashes.size() + 1) & ~1ULL); // capacity rounded up to even
```
This reminds me of the [Obfuscated C Code Contest](https://github.com/ioccc-src/winner/blob/master/2024/burton/prog.c). I consider the dumb version easier to read:

```cpp
x % 2 == 0 ? x : x + 1
```
Remember that other people could get confused easier than you.

l0rinc replied:
> maybe put it in an inline function RoundUpToEven(n)

I did that before, reviewers found it complicated.

vasild commented on src/bench/merkle_root.cpp:
```diff
@@ -7,21 +7,34 @@
+    FastRandomContext rng{/*fDeterministic=*/true};
+
+    std::vector<uint256> hashes{};
+    hashes.resize(9001);
```
In the commit message of 8af0a3a72c8af781bebed7289657c637220df893 `validation: pre-reserve leaves to prevent reallocs with odd vtx count`:

> ... doubling its capacity (allocating 3x the necessary memory)

Why 3x?

vasild reviewed (commented):
Approach ACK 8af0a3a72c8af781bebed7289657c637220df893

hodlinator reviewed (approved):
re-ACK 9f39a8309cb0a23df03937b0b225aff8ce9e1ec6

Resolved conflict with #33805 and fixed code nit https://github.com/bitcoin/bitcoin/pull/32497#discussion_r2602163619.

vasild reviewed (approved):
ACK 9f39a8309cb0a23df03937b0b225aff8ce9e1ec6

hodlinator reviewed (approved):
re-ACK 3dd815f048c80c9a35f01972e0537eb42531aec7

Changes since previous ACK (https://github.com/bitcoin/bitcoin/pull/32497#pullrequestreview-3567608553):
* Remove confusing assert in benchmark (https://github.com/bitcoin/bitcoin/pull/32497#discussion_r2641466404).
* Updated benchmark timing results in commit messsages.

vasild reviewed (approved):
ACK 3dd815f048c80c9a35f01972e0537eb42531aec7

w0xlt commented on src/test/merkle_tests.cpp:
```diff
@@ -242,12 +243,12 @@ BOOST_AUTO_TEST_CASE(merkle_test_BlockWitness)
-    hashes[0].SetNull();
-    hashes[1] = block.vtx[1]->GetHash().ToUint256();
+    hashes.resize(vtx_count); // Note: leaving odd count to exercise old behavior
+    for (size_t pos{1}; pos < vtx_count; ++pos) {
```
nit: The old code explicitly had `hashes[0].SetNull()`. Now `hashes[0]` relies on implicit zero-initialization from `resize()`. A brief comment would clarify intent:
```suggestion
    // hashes[0] is zero (coinbase witness hash) from resize()
    for (size_t pos{1}; pos < vtx_count; ++pos) {
```

l0rinc replied:
I can do that in a follow-up if you think it's necessarry

w0xlt reviewed (commented):
ACK 3dd815f048c80c9a35f01972e0537eb42531aec7 with minor nits.

On my machine, the Massif data clearly shows the improvement.

Benchmark Results
  
| Version | Benchmark | ns/leaf | leaf/s |
| ------------- | ------------- |  ------------- |  ------------- |
| After (optimized)| MerkleRoot |46.50 | 21.5M |
| After (optimized) | MerkleRootWithMutation | 45.97 | 21.8M |
| Before (master) | MerkleRoot | 47.19 | 21.2M |

  Massif Memory Results
  
| Metric | Before (master) | After (optimized) | Change |
| ------------- | ------------- |  ------------- |  ------------- |
| Peak memory | 1.34 MB | 802 KB | -40% |
| Reallocation overhead | 576 KB (41%) | 0 | Eliminated |

1. Memory reduction: 1.34 MB → 0.80 MB
2. Reallocation eliminated: The _M_realloc_insert stack trace is gone entirely
3. No performance regression: ~47 ns/leaf before and after

l0rinc commented on src/test/fuzz/integer.cpp:
```diff
@@ -80,8 +80,8 @@ FUZZ_TARGET(integer, .init = initialize_integer)
     constexpr uint256 u256_max{"ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"};
-    const std::vector<uint256> v256{u256, u256_min, u256_max};
-    (void)ComputeMerkleRoot(v256);
+    std::vector v256{u256, u256_min, u256_max};
```
This was deliberately removed since it's redundant

l0rinc commented on src/bench/merkle_root.cpp:
```diff
@@ -7,21 +7,33 @@
+            leaves.reserve((hashes.size() + 1) & ~1ULL); // capacity rounded up to even
+            for (size_t s = 0; s < hashes.size(); s++) {
+                leaves.push_back(hashes[s]);
+            }
```
This was deliberately left to match the production code, see https://github.com/bitcoin/bitcoin/pull/32497#discussion_r2568112479

danielabrozzoni reviewed (commented):
Code review ACK 3dd815f048

l0rinc commented on src/test/merkle_tests.cpp:
```diff
@@ -242,12 +243,12 @@ BOOST_AUTO_TEST_CASE(merkle_test_BlockWitness)
-    hashes.resize(block.vtx.size());
-    hashes[0].SetNull();
-    hashes[1] = block.vtx[1]->GetHash().ToUint256();
+    hashes.resize(vtx_count); // Note: leaving odd count to exercise old behavior
```
Done in https://github.com/bitcoin/bitcoin/pull/34376
